core/web/evm_transfer_controller.go

```diff
@@ -94,26 +93,13 @@ func ValidateEthBalanceForTransfer(c *gin.Context, chain evm.Chain, fromAddr com
 		return errors.Errorf("balance is too low for this transaction to be executed: %v", balance)
 	}
 
-	var fees gas.EvmFee
-
 	gasLimit := chain.Config().EVM().GasEstimator().LimitTransfer()
 	estimator := chain.GasEstimator()
 
-	fees, gasLimit, err = estimator.GetFee(c, nil, gasLimit, chain.Config().EVM().GasEstimator().PriceMaxKey(fromAddr))
+	amountWithFees, err := estimator.GetMaxCost(c, amount, nil, gasLimit, chain.Config().EVM().GasEstimator().PriceMaxKey(fromAddr))
```

Comment:
Personally, I don't see a good reason to extend the existing EvmFeeEstimator interface with another method that will only be used by one caller. As Sam suggested above, this method doesn't exactly return the actual max cost, so it might not be that reusable by other callers. `ValidateEthBalanceForTransfer` already encapsulates the logic, so I would rather implement the logic here instead:

```suggestion
	fees, gasLimit, err = estimator.GetFee(c, nil, gasLimit, chain.Config().EVM().GasEstimator().PriceMaxKey(fromAddr))
	if err != nil {
		return errors.Wrap(err, "failed to estimate gas")
	}

	var gasPrice *assets.Wei
	if chain.Config().EVM().GasEstimator().EIP1559DynamicFees() {
		gasPrice = fees.DynamicFeeCap
	} else {
		gasPrice = fees.Legacy
	}
```

Reply:
+1 move it

Reply:
> with another method that will only be used by one caller

@dimriou The point is that EIP-1559 logic details do not belong at this level of the API hierarchy. Reusability isn't the concern here, it's the principle of least surprise. 

Why would returning both a legacy and dynamic fee make sense? Are there multiple callers throughout the code base that need to compare the difference between a legacy and a dynamic fee?

Reply:
another option could be to move the calculation logic as a utility function rather than part of the estimator interface?

that way we can do both: separate the calculation from the API level and not require it in the base estimator interface?

Reply:
A utility function is better than the raw implementation, but isn't it still true that each caller of `GetFee` has to make a decision based on whether EIP-1559 is enabled or not?

If we rename the function to what it's doing: `GetLegacyAndEIP1559Fee` is it more clear on the burden this API gives to its callers? 

Another option: 
Should chain.GasEstimator() return a LegacyWrappedEvmEstimator if EIP-1559 is not enabled and a WrappedEvmEstimator if it is enabled? This seemed like a deeper refactor so I didn't suggest it previously.

Reply:
`GetMaxCost()` still doesn't make sense to be part of the API as it abstracts 4 lines of code that are only usable by this controller. We've worked hard to make the API simpler, and this goes in the opposite direction.

> Should chain.GasEstimator() return a LegacyWrappedEvmEstimator if EIP-1559 is not enabled and a WrappedEvmEstimator if it is enabled? This seemed like a deeper refactor so I didn't suggest it previously.

I think the problem you mention is with the fee estimator and specifically how `GetFee()` method abstracts fees between dynamic and legacy transactions. The above suggestion seems far better TBH and solves both problems, although since it requires a deeper refactor and there is consensus, I'm ok to have it as it is.

Reply:
> solves both problems

We should do it then! This PR unblocks 2.4.0, so there is no rush on implementing the better solution now.
cc @aalu1418 if you are interested, but I can also take this on.